python/pyspark/sql/dataframe.py

```diff
@@ -1648,8 +1650,30 @@ def toPandas(self):
         0    2  Alice
         1    5    Bob
         """
-        import pandas as pd
-        return pd.DataFrame.from_records(self.collect(), columns=self.columns)
+        if self.sql_ctx.getConf("spark.sql.execution.arrow.enable", "false").lower() == "true":
+            try:
+                import pyarrow
+                tables = self._collectAsArrow()
+                table = pyarrow.concat_tables(tables)
```

Comment:
If tables is an empty list (e.g. if you load a dataset, filter the whole thing, and produce zero rows), `pyarrow.concat_tables` raises an exception rather than producing an empty table.  This should probably be fixed in arrow (cc @wesm) but we should be defensive here.  Probably should try to produce a `DataFrame` with the right schema but no rows if possible.